llvm/tools/sycl-post-link/CompileTimePropertiesPass.cpp

```diff
@@ -142,9 +152,124 @@ PreservedAnalyses CompileTimePropertiesPass::run(Module &M,
+              dyn_cast<ConstantAggregate>(AnnotValsGV->getInitializer())) {
+        assert(
+            (AnnotValsAggr->getNumOperands() & 1) == 0 &&
+            "sycl-properties annotation must have an even number of annotation "
+            "values.");
+
+        // Iterate over the pairs of property meta-names and meta-values.
+        for (size_t I = 0; I < AnnotValsAggr->getNumOperands(); I += 2) {
+          Optional<StringRef> PropMetaName =
+              getGlobalVariableString(AnnotValsAggr->getOperand(I));
+          Optional<StringRef> PropMetaValue =
+              getGlobalVariableString(AnnotValsAggr->getOperand(I + 1));
+
+          assert(PropMetaName &&
+                 "Unexpected format for property name in annotation.");
+
+          auto DecorIt = SpirvDecorMap.find(*PropMetaName);
+          if (DecorIt == SpirvDecorMap.end())
+            continue;
+          uint32_t DecorCode = DecorIt->second.Code;
+
+          // Expected format is '{X}' or '{X:Y}' where X is decoration ID and
+          // Y is the value if present. It encloses Y in " to ensure that
+          // string values are handled correctly. Note that " around values are
+          // always valid, even if the decoration parameters are not strings.
+          NewAnnotString += "{" + std::to_string(DecorCode);
+          if (PropMetaValue)
+            NewAnnotString += ":\"" + PropMetaValue->str() + "\"";
+          NewAnnotString += "}";
+        }
+      }
+    }
+  }
+
+  // If the new annotation string is empty there is no reason to keep it, so
+  // replace it with the first operand and mark it for removal.
+  if (NewAnnotString == "") {
+    IntrInst->replaceAllUsesWith(IntrInst->getOperand(0));
+    RemovableAnnotations.push_back(IntrInst);
+    return true;
+  }
+
+  // Either reuse a previously generated one or create a new global variable
+  // with the new annotation string.
+  GlobalVariable *NewAnnotStringGV = nullptr;
+  auto ExistingNewAnnotStringIt = NewAnnotationStrings.find(NewAnnotString);
+  if (ExistingNewAnnotStringIt != NewAnnotationStrings.end()) {
+    NewAnnotStringGV = ExistingNewAnnotStringIt->second;
+  } else {
+    Constant *NewAnnotStringData =
+        ConstantDataArray::getString(M.getContext(), NewAnnotString);
+    NewAnnotStringGV = new GlobalVariable(M, NewAnnotStringData->getType(),
+                                          true, GlobalValue::PrivateLinkage,
+                                          NewAnnotStringData, ".str");
+    NewAnnotStringGV->setSection(AnnotStrArgGV->getSection());
+    NewAnnotStringGV->setUnnamedAddr(GlobalValue::UnnamedAddr::Global);
+    NewAnnotationStrings.insert({NewAnnotString, NewAnnotStringGV});
+  }
+
+  // Replace the annotation string with a bitcast of the new global variable.
+  Type *Int8Ty = IntegerType::getInt8Ty(M.getContext());
+  IntrInst->setArgOperand(
+      1, ConstantExpr::getBitCast(NewAnnotStringGV, Int8Ty->getPointerTo(0)));
```

Comment:
Does `0` mean the private address space?

Reply:
I believe so, but it was taken from the behavior of typical annotation intrinsic codegen. I've changed it to take the type from the argument it is replacing.